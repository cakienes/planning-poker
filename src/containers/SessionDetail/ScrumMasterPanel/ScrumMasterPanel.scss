.scrumMasterPanel {
    margin-top: 52px;
    .storyDetail {
        border: 1px solid #000;
        padding-right: 30px;
        position: relative;
        margin-top: 20px;
        .panelName {
            position: absolute;
            top: -13px;
            background: #fff;
            left: 15px;
        }
        .detailBox {
            margin-top: 20px;
            padding: 10px;
            .storyName {
            }
            .voterList {
                margin-top: 20px;
                ul {
                    list-style: none;
                    margin: 0;
                    padding: 0;
                }
            }
        }
    }
}
