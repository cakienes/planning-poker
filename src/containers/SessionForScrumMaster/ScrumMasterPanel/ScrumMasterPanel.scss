.scrumMasterPanel {
    margin-top: 52px;
    .storyDetail {
        border: 1px solid #000;
        position: relative;
        margin-top: 20px;
        padding-bottom: 30px;
        .panelName {
            position: absolute;
            top: -13px;
            background: #fff;
            left: 15px;
        }
        .detailBox {
            margin-top: 20px;
            padding: 10px;
            .storyName {
            }
            .voterList {
                margin-top: 20px;
                ul {
                    list-style: none;
                    margin: 0;
                    padding: 0;
                }
            }
        }
    }
}
