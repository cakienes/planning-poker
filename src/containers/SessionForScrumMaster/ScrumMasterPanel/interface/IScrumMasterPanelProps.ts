import ISession from '../../../../interfaces/ISession';
import IUserStory from '../../../../interfaces/IUserStory';

export default interface IScrumMasterPanelProps {
    activeUserStory?: IUserStory;
    selectedSession?: ISession;
    setFinalScore: (finalScore: string) => void;
}
