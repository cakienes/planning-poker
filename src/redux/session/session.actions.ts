import ISession from '../../interfaces/ISession';
import {
    ADD_SESSION,
    GET_SESSIONS_FROM_LOCAL_STORAGE,
    GIVE_STORY_POINT,
    SET_DEVELOPER,
    SET_FINAL_SCORE,
    SET_NEW_DEVELOPER,
    SET_SELECTED_SESSION,
    SET_SELECTED_SESSION_BY_URL,
} from './session.constants';

export const getSessionsFromLocalStorage = () => (dispatch: Function) => {
    dispatch({
        type: GET_SESSIONS_FROM_LOCAL_STORAGE,
    });
};

export const createSession = (session: ISession) => (dispatch: Function) => {
    dispatch({
        type: ADD_SESSION,
        session,
    });
};
export const setSelectedSession = (id: string) => (dispatch: Function) => {
    dispatch({
        type: SET_SELECTED_SESSION,
        id,
    });
};

export const giveStoryPoint = (voterName: string, storyPoint?: number) => (dispatch: Function) => {
    dispatch({
        type: GIVE_STORY_POINT,
        voterName,
        storyPoint,
    });
};

export const setNewDeveloper = (guid: string) => (dispatch: Function) => {
    dispatch({
        type: SET_NEW_DEVELOPER,
        guid,
    });
};
export const setDeveloper = (developerId: string) => (dispatch: Function) => {
    dispatch({
        type: SET_DEVELOPER,
        developerId,
    });
};
export const setFinalScore = (finalScore: string) => (dispatch: Function) => {
    dispatch({
        type: SET_FINAL_SCORE,
        finalScore,
    });
};
export const setSelecedSessionByUrl = (optionalUrl: string) => (dispatch: Function) => {
    dispatch({
        type: SET_SELECTED_SESSION_BY_URL,
        optionalUrl,
    });
};
