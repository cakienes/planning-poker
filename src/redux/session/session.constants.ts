export const GET_SESSIONS_FROM_LOCAL_STORAGE = 'GET_SESSIONS_FROM_LOCAL_STORAGE';
export const ADD_SESSION = 'ADD_SESSION';
export const SET_SELECTED_SESSION = 'SET_SELECTED_SESSION';
export const GIVE_STORY_POINT = 'GIVE_STORY_POINT';
export const SET_NEW_DEVELOPER = 'SET_NEW_DEVELOPER';
export const SET_DEVELOPER = 'SET_DEVELOPER';
