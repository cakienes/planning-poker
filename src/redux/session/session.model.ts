import ISession from '../../interfaces/ISession';

export default interface ISessionModel {
    sessions?: ISession[];
    selectedSessionId?: string;
    developerId?: string;
}
