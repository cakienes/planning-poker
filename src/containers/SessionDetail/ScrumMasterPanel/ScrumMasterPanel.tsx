import React from 'react';
import { connect } from 'react-redux';
import IGlobalState from '../../../interfaces/IGlobalState';
import IVoter from '../../../interfaces/IVoter';
import { getActiveUserStory, getSelectedSession } from '../../../redux/session/session.selectors';
import IScrumMasterPanelProps from './interface/IScrumMasterPanelProps';
import './ScrumMasterPanel.scss';
export class ScrumMasterPanel extends React.Component<IScrumMasterPanelProps, any> {
    render() {
        const { activeUserStory } = this.props;
        return (
            <div className="scrumMasterPanel">
                <div className="storyDetail">
                    <div className="panelName">Scrum Master Panel</div>
                    <div className="detailBox">
                        <div className="storyName">{activeUserStory && activeUserStory.storyName}</div>
                        <div className="voterList">{this.renderVoters()}</div>
                    </div>
                </div>
            </div>
        );
    }

    renderVoters = (): React.ReactNode => {
        const { activeUserStory, selectedSession } = this.props;
        if (selectedSession && activeUserStory) {
            const array: string[] = [];
            for (let index = 1; index < Number(selectedSession.numberOfVoters) + 1; index++) {
                array.push(`Voter-${index}`);
            }
            return (
                <ul>
                    {array.map((voter: string) => (
                        <li key={voter}>
                            {voter} : {this.getVoteNumber(voter)}
                        </li>
                    ))}
                    <li>Scrum Master : 22</li>
                </ul>
            );
        }
    };

    getVoteNumber = (voterName: string): string => {
        const { activeUserStory } = this.props;
        if (activeUserStory && activeUserStory.voters) {
            const voter: IVoter | undefined = activeUserStory.voters.find(x => x.voterName === voterName.toUpperCase());
            if (voter && voter.storyPoint !== 1000) return voter.storyPoint.toString();
            else if (voter && voter.storyPoint === 1000) return '?';
        }
        return 'Not Voted';
    };
}

const mapStateToProps = (state: IGlobalState) => ({
    activeUserStory: getActiveUserStory(state),
    selectedSession: getSelectedSession(state),
});

export default connect(
    mapStateToProps,
    null,
)(ScrumMasterPanel);
