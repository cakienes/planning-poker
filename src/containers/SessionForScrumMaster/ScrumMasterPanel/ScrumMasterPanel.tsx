import React from 'react';
import { connect } from 'react-redux';
import { UserTypeEnum } from '../../../helper/Enum';
import IGlobalState from '../../../interfaces/IGlobalState';
import IVoter from '../../../interfaces/IVoter';
import { setFinalScore } from '../../../redux/session/session.actions';
import { getActiveUserStory, getSelectedSession } from '../../../redux/session/session.selectors';
import FinalScoreForm from './FinalScoreForm/FinalScoreForm';
import IScrumMasterPanelProps from './interface/IScrumMasterPanelProps';
import './ScrumMasterPanel.scss';
export class ScrumMasterPanel extends React.Component<IScrumMasterPanelProps, any> {
    render() {
        const { activeUserStory } = this.props;
        return (
            <div className="scrumMasterPanel">
                <div className="storyDetail">
                    <div className="panelName">Scrum Master Panel</div>
                    <div className="detailBox">
                        <div className="storyName">{activeUserStory && activeUserStory.storyName}</div>
                        <div className="voterList">{this.renderVoters()}</div>
                    </div>
                    <div className="text-center">{this.renderDownBlock()}</div>
                </div>
            </div>
        );
    }

    // TODO Refactor if have time
    renderVoters = (): React.ReactNode => {
        const { activeUserStory, selectedSession } = this.props;
        if (selectedSession && activeUserStory) {
            return (
                <ul>
                    {selectedSession.developers.map((developerId: string, index: number) => (
                        <li key={developerId}>
                            Voter-{index + 1} : {this.getVoteNumber(developerId)}
                        </li>
                    ))}
                    <li>Scrum Master : {this.getVoteNumber(UserTypeEnum.SCRUM_MASTER)}</li>
                </ul>
            );
        }
    };

    getVoteNumber = (developerId: string): string => {
        const { activeUserStory, selectedSession } = this.props;
        if (activeUserStory && activeUserStory.voters && selectedSession) {
            const voter: IVoter | undefined = activeUserStory.voters.find(x => x.voterName === developerId);
            if (activeUserStory.voters.length === Number(selectedSession.numberOfVoters) + 1) {
                if (voter && voter.storyPoint) return voter.storyPoint.toString();
                else if (voter && !voter.storyPoint) return '?';
            }
            if (voter) return 'Voted';
        }
        return 'Not Voted';
    };

    renderDownBlock = (): React.ReactNode => {
        const { activeUserStory, selectedSession } = this.props;
        if (activeUserStory) {
            if (
                selectedSession &&
                activeUserStory.voters &&
                activeUserStory.voters.length === Number(selectedSession.numberOfVoters) + 1
            ) {
                return <FinalScoreForm onSubmit={this.setFinalScore} activeUserStory={activeUserStory} />;
            }
            return <div>You can not end voting till each teammate voted</div>;
        }
        return <div>SESSION ENDED</div>;
    };

    setFinalScore = (values: any): void => {
        this.props.setFinalScore(values.finalScore);
    };
}

const mapStateToProps = (state: IGlobalState) => ({
    activeUserStory: getActiveUserStory(state),
    selectedSession: getSelectedSession(state),
});

export const mapDispatchToProps = (dispatch: Function) => {
    return {
        setFinalScore: (finalScore: string) => {
            dispatch(setFinalScore(finalScore));
        },
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(ScrumMasterPanel);
