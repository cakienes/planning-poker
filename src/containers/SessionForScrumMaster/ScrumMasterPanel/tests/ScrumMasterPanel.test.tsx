import React from 'react';
import { createRenderer, ShallowRenderer } from 'react-test-renderer/shallow';
import { setFinalScore } from '../../../../redux/session/session.actions';
import IScrumMasterPanelProps from '../interface/IScrumMasterPanelProps';
import { mapDispatchToProps, ScrumMasterPanel } from '../ScrumMasterPanel';

describe('ScrumMasterPanel.tsx', () => {
    let renderer: ShallowRenderer;
    const setFinalScoreMock = jest.fn();

    let scrumMasterPanelProps: IScrumMasterPanelProps = {} as any;

    beforeEach(() => {
        renderer = createRenderer();
        scrumMasterPanelProps = {
            setFinalScore: setFinalScoreMock,
            activeUserStory: undefined,
            selectedSession: undefined,
        };
    });

    it('renders correctly ScrumMasterPanel', async () => {
        expect(renderer.render(<ScrumMasterPanel {...scrumMasterPanelProps} />)).toMatchSnapshot();
    });

    it('renders correctly ScrumMasterPanel while activeUserStory is defined', async () => {
        scrumMasterPanelProps = {
            ...scrumMasterPanelProps,
            activeUserStory: {
                storyName: 'storyName',
            },
        };
        expect(renderer.render(<ScrumMasterPanel {...scrumMasterPanelProps} />)).toMatchSnapshot();
    });

    it('renderVoters of ScrumMasterPanel while selectedSession && activeUserStory is defined', async () => {
        scrumMasterPanelProps = {
            ...scrumMasterPanelProps,
            activeUserStory: {
                storyName: 'storyName',
            },
            selectedSession: {
                developers: ['developerId', 'developerId2'],
                id: 'id',
                numberOfVoters: 2,
                optionalUrl: 'optionalUrl',
                sessionName: 'sessionName',
                userStories: [
                    {
                        storyName: 'storyName',
                    },
                ],
            },
        };
        renderer.render(<ScrumMasterPanel {...scrumMasterPanelProps} />);
        const instance: any = renderer.getMountedInstance();

        expect(instance.renderVoters()).toMatchSnapshot();
    });

    it('getVoteNumber of ScrumMasterPanel while VOTED', async () => {
        scrumMasterPanelProps = {
            ...scrumMasterPanelProps,
            activeUserStory: {
                storyName: 'storyName',
                voters: [{ voterName: 'developerId', storyPoint: 5 }],
            },
            selectedSession: {
                developers: ['developerId', 'developerId2'],
                id: 'id',
                numberOfVoters: 2,
                optionalUrl: 'optionalUrl',
                sessionName: 'sessionName',
                userStories: [
                    {
                        storyName: 'storyName',
                        voters: [{ voterName: 'developerId', storyPoint: 5 }],
                    },
                ],
            },
        };
        renderer.render(<ScrumMasterPanel {...scrumMasterPanelProps} />);
        const instance: any = renderer.getMountedInstance();

        expect(instance.getVoteNumber('developerId')).toEqual('Voted');
    });

    it('getVoteNumber of ScrumMasterPanel while NOT_VOTED', async () => {
        scrumMasterPanelProps = {
            ...scrumMasterPanelProps,
            activeUserStory: {
                storyName: 'storyName',
                voters: [{ voterName: 'developerId', storyPoint: 5 }],
            },
            selectedSession: {
                developers: ['developerId', 'developerId2'],
                id: 'id',
                numberOfVoters: 2,
                optionalUrl: 'optionalUrl',
                sessionName: 'sessionName',
                userStories: [
                    {
                        storyName: 'storyName',
                        voters: [{ voterName: 'developerId', storyPoint: 5 }],
                    },
                ],
            },
        };
        renderer.render(<ScrumMasterPanel {...scrumMasterPanelProps} />);
        const instance: any = renderer.getMountedInstance();

        expect(instance.getVoteNumber('developerId2')).toEqual('Not Voted');
    });

    it('getVoteNumber of ScrumMasterPanel while all VOTED', async () => {
        scrumMasterPanelProps = {
            ...scrumMasterPanelProps,
            activeUserStory: {
                storyName: 'storyName',
                voters: [
                    { voterName: 'developerId', storyPoint: 0 },
                    { voterName: 'developerId2', storyPoint: 5 },
                    { voterName: 'SCRUM_MASTER', storyPoint: 5 },
                ],
            },
            selectedSession: {
                developers: ['developerId', 'developerId2'],
                id: 'id',
                numberOfVoters: 2,
                optionalUrl: 'optionalUrl',
                sessionName: 'sessionName',
                userStories: [
                    {
                        storyName: 'storyName',
                        voters: [
                            { voterName: 'developerId', storyPoint: 0 },
                            { voterName: 'developerId2', storyPoint: 5 },
                            { voterName: 'SCRUM_MASTER', storyPoint: 5 },
                        ],
                    },
                ],
            },
        };
        renderer.render(<ScrumMasterPanel {...scrumMasterPanelProps} />);
        const instance: any = renderer.getMountedInstance();

        expect(instance.getVoteNumber('developerId')).toEqual('?');
        expect(instance.getVoteNumber('SCRUM_MASTER')).toEqual('5');
    });

    it('setFinalScore of ScrumMasterPanel ', async () => {
        renderer.render(<ScrumMasterPanel {...scrumMasterPanelProps} />);
        const instance: any = renderer.getMountedInstance();

        const dumValues = {
            finalScore: '5',
        };
        instance.setFinalScore(dumValues);

        expect(setFinalScoreMock).toHaveBeenCalledWith('5');
    });

    describe('mapDispatchToProps', () => {
        it('should dispatch getSessionsFromLocalStorage when called', () => {
            const dispatchMock = jest.fn();
            const result = mapDispatchToProps(dispatchMock);
            result.setFinalScore('13');
            expect(JSON.stringify(dispatchMock.mock.calls[0][0])).toEqual(JSON.stringify(setFinalScore('13')));
        });
    });
});
