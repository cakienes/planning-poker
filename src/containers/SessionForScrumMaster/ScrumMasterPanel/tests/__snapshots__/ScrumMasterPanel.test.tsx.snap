// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrumMasterPanel.tsx renderVoters of ScrumMasterPanel while selectedSession && activeUserStory is defined 1`] = `
<ul>
  <li>
    Voter-
    1
     : 
    Not Voted
  </li>
  <li>
    Voter-
    2
     : 
    Not Voted
  </li>
  <li>
    Scrum Master : 
    Not Voted
  </li>
</ul>
`;

exports[`ScrumMasterPanel.tsx renders correctly ScrumMasterPanel 1`] = `
<div
  className="scrumMasterPanel"
>
  <div
    className="storyDetail"
  >
    <div
      className="panelName"
    >
      Scrum Master Panel
    </div>
    <div
      className="detailBox"
    >
      <div
        className="storyName"
      />
      <div
        className="voterList"
      />
    </div>
    <div
      className="text-center"
    >
      <div>
        SESSION ENDED
      </div>
    </div>
  </div>
</div>
`;

exports[`ScrumMasterPanel.tsx renders correctly ScrumMasterPanel while activeUserStory is defined 1`] = `
<div
  className="scrumMasterPanel"
>
  <div
    className="storyDetail"
  >
    <div
      className="panelName"
    >
      Scrum Master Panel
    </div>
    <div
      className="detailBox"
    >
      <div
        className="storyName"
      >
        storyName
      </div>
      <div
        className="voterList"
      />
    </div>
    <div
      className="text-center"
    >
      <div>
        You can not end voting till each teammate voted
      </div>
    </div>
  </div>
</div>
`;
